Make all integer intrinsics generic

GBGamer opened the pull request:
Similarly to the simd intrinsics. I believe this is a better solution than #29288, and I could implement it as well for overflowing_add/sub/mul. Also rename from udiv/sdiv to div, and same for rem.

rust-highfive commented:
r? @alexcrichton

(rust_highfive has picked a reviewer for you, use r? to override)

alexcrichton commented on src/libcore/intrinsics.rs:
```diff
@@ -609,3 +595,14 @@ extern "rust-intrinsic" {
+extern "rust-intrinsic" {
+    /// Performs an unchecked division, resulting in undefined behavior
+    /// where y = 0 or x = `T::min_value()` and y = -1
+    pub fn unchecked_div<T>(x: T, y: T) -> T;
```
I think these can be in the same place they were in above, the `#[cfg]` should work ok on the functions as well as the block itself.

alexcrichton commented on src/librustc_trans/trans/intrinsic.rs:
```diff
@@ -792,10 +792,33 @@ pub fn trans_intrinsic_call<'a, 'blk, 'tcx>(mut bcx: Block<'blk, 'tcx>,
-        (_, "unchecked_sdiv") => SDiv(bcx, llargs[0], llargs[1], call_debug_location),
-        (_, "unchecked_urem") => URem(bcx, llargs[0], llargs[1], call_debug_location),
-        (_, "unchecked_srem") => SRem(bcx, llargs[0], llargs[1], call_debug_location),
+        (_, c) if c == "unchecked_div" || c == "unchecked_rem" => {
```
Perhaps this could be:

```rust
(_, c @ "unchecked_div") | (_, c @ "unchecked_rem") => { /* ... * }
```

(avoids the `if` guard

alexcrichton commented on src/librustc_trans/trans/intrinsic.rs:
```diff
@@ -792,10 +792,33 @@ pub fn trans_intrinsic_call<'a, 'blk, 'tcx>(mut bcx: Block<'blk, 'tcx>,
+                } else if c == "unchecked_rem" {
+                    SRem(bcx, llargs[0], llargs[1], call_debug_location)
+                } else {
+                    unreachable!()
```
I think it's fine to just have the `else` clause fall down to `SRem` and not worry about the `unreachable!` here

GBGamer commented:
@alexcrichton alright, I made the changes you suggested.

alexcrichton commented:
Thanks! Could you also squash to one commit now? Otherwise r=me

GBGamer commented:
@alexcrichton this has blown up far more. I've replaced all integer intrinsics with generics, and do this same thing with all of them (And added two errors in the process, one for calling bswap with u/i8, and the other is the error you see in this PR).

alexcrichton commented on src/librustc_trans/trans/intrinsic.rs:
```diff
@@ -589,217 +582,115 @@ pub fn trans_intrinsic_call<'a, 'blk, 'tcx>(mut bcx: Block<'blk, 'tcx>,
+                    ast::TyU64 => 64,
+                };
+                match c {
+                    "ctlz" => count_zeros_intrinsic(bcx, &format!("llvm.ctlz.i{}", width),
```
It seems like the int/uint branches here share quite a bit of code, perhaps they could be deduplicated?

GBGamer replied:
Unfortunately no, as they are two different types. The most I could do is unduplicate the target_pointer_width part.

alexcrichton replied:
Hm are you sure? It looks like everything is basically duplicated except for unsigned/signed in a few locations which could be a parameter to a shared function? For example the bswap intrinsic has exactly the same code in both branches here.

alexcrichton commented:
@GBGamer looks good to me! I'll get the second opinion of someone else on the compilers team, however, as it's probably good to run this past others as well.

cc @rust-lang/compiler 

r? @nrc (or anyone else really)

eddyb commented on src/libcore/num/mod.rs:
```diff
@@ -611,54 +612,110 @@ macro_rules! int_impl {
+    int_impl! { i8, u8, 8,
+        intrinsics::add_with_overflow,
+        intrinsics::sub_with_overflow,
+        intrinsics::mul_with_overflow }
```
Since these are all the same, can't the macro just not take them anymore when `#[cfg(not(stage0))]`?

GBGamer replied:
Yes, but the macro is 500 lines long. It's better to add it here until we get these into stage0, and then change the macro.

eddyb replied:
Did not realize that, makes sense.

eddyb commented on src/libcore/num/mod.rs:
```diff
@@ -1173,8 +1251,21 @@ impl u8 {
+        intrinsics::add_with_overflow,
+        intrinsics::sub_with_overflow,
+        intrinsics::mul_with_overflow }
+}
```
Same here.

GBGamer replied:
Same as above. This is ~50 lines, versus 500 lines, duplicated.

eddyb commented on src/libcore/num/wrapping.rs:
```diff
@@ -12,23 +12,42 @@
+}
+
+
+use self::staged_uses::*;
```
Doesn't `#[cfg(...)]` work on `use` statements?

GBGamer replied:
Yes it does, I did not know that.

eddyb commented on src/librustc_trans/trans/intrinsic.rs:
```diff
@@ -589,217 +582,86 @@ pub fn trans_intrinsic_call<'a, 'blk, 'tcx>(mut bcx: Block<'blk, 'tcx>,
+        (_, c @ "ctlz") | (_, c @ "cttz") | (_, c @ "ctpop") | (_, c @ "bswap") |
+        (_, c @ "add_with_overflow") | (_, c @ "sub_with_overflow") | (_, c @ "mul_with_overflow") |
+        (_, c @ "overflowing_add") | (_, c @ "overflowing_sub") | (_, c @ "overflowing_mul") |
+        (_, c @ "unchecked_div") | (_, c @ "unchecked_rem") => {
```
You don't need all the `c @ ...` because `c` is already in the `name` variabile.

GBGamer replied:
Oh. That makes it much nicer.

eddyb commented on src/librustc_trans/trans/intrinsic.rs:
```diff
@@ -589,217 +582,86 @@ pub fn trans_intrinsic_call<'a, 'blk, 'tcx>(mut bcx: Block<'blk, 'tcx>,
+                                with_overflow_intrinsic(
+                                    bcx, &format!("llvm.umul.with.overflow.i{}", width),
+                                    llargs[0], llargs[1], llresult, call_debug_location)
+                            },
```
All of these could use the same implementation, with "s" or "u" decided by `signed` and the operation name from `name[..3]`.

GBGamer replied:
Thank you for this, this makes it so much nicer :)

eddyb commented on src/librustc_trans/trans/intrinsic.rs:
```diff
@@ -1716,3 +1578,39 @@ fn generic_simd_intrinsic<'blk, 'tcx, 'a>
 }
+
+// Returns the width of an int TypeVariant, and if it's signed or not
+// Returns None if the type is not an integer
```
There is nothing like this anywhere already?

eddyb replied:
I just checked, you can call `t.bit_width()` below, but that doesn't help with `TyIs` and `TyUs`.
I'm surprised, but I guess your solution is good (you could put this somewhere more general, like a `Ty` method, though).

eddyb commented on src/librustc_typeck/check/intrinsic.rs:
```diff
@@ -254,60 +254,22 @@ pub fn check_intrinsic_type(ccx: &CrateCtxt, it: &hir::ForeignItem) {
+            "ctpop"      => (1, vec!(param(ccx, 0)), param(ccx, 0)),
+            "ctlz"       => (1, vec!(param(ccx, 0)), param(ccx, 0)),
+            "cttz"       => (1, vec!(param(ccx, 0)), param(ccx, 0)),
+            "bswap"      => (1, vec!(param(ccx, 0)), param(ccx, 0)),
```
Could group these into one match arm.

GBGamer replied:
Yup.

eddyb commented:
LGTM modulo nits.

GBGamer commented:
Alright, I think I've covered all your nits @eddyb?

eddyb commented:
@bors r+

bors commented:
:pushpin: Commit 29808c4 has been approved by `eddyb`

<!-- @bors r=eddyb 29808c421da00d89b7749a6bd19c8c9fa30ae7d0 -->

bors commented:
:hourglass: Testing commit 29808c4 with merge 10a4130...

bors commented:
:broken_heart: Test failed - [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/6904)

GBGamer commented:
Blargh, fixing.

GBGamer commented:
hopefully fixed :'(

eddyb commented:
@bors r+

bors commented:
:pushpin: Commit 93ca0fa has been approved by `eddyb`

<!-- @bors r=eddyb 93ca0fac9e4704f0ecf2d21e29a6f97a8eb7f6c1 -->

bors commented:
:hourglass: Testing commit 93ca0fa with merge 99ac88e...

bors commented:
:broken_heart: Test failed - [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/6975)

GBGamer commented:
alright should be fixed for reals this time :'(

eddyb commented:
@bors r+

bors commented:
:pushpin: Commit 579420f has been approved by `eddyb`

<!-- @bors r=eddyb 579420fbdd9951ae230a9def03e157d9b9957b2f -->

bors commented:
:hourglass: Testing commit 579420f with merge a5fbb3a...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/6950), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/6964), [auto-linux-64-debug-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/223), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/6903), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/6926), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/6909), [auto-linux-cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/144), [auto-linux-musl-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/2094), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/6979), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/6975), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/6940), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/1931), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/1932), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/1933), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/1931), [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/1195), [auto-win-msvc-64-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/1915)
